LK-MPI documentation update (#1190)
PennyLaneAI/pennylane-lightning: doc/lightning_kokkos/installation.rst

.. code-block:: bash

    git clone https://github.com/PennyLaneAI/pennylane-lightning.git
    cd pennylane-lightning
    PL_BACKEND="lightning_qubit" python scripts/configure_pyproject_toml.py
    SKIP_COMPILATION=True pip install -e . --config-settings editable_mode=compat
    PL_BACKEND="lightning_kokkos" python scripts/configure_pyproject_toml.py
    CMAKE_ARGS="-DENABLE_MPI=ON" python -m pip install -e . --config-settings editable_mode=compat -vv

If required, extra linker flags for MPI (e.g. for GPU Transport Layer) can be added using the ``MPI_EXTRA_LINKER_FLAGS`` environment variable, for example:

.. code-block:: bash

    export MPI_EXTRA_LINKER_FLAGS="-lxpmem -L/opt/cray/pe/mpich/8.1.31/gtl/lib -lmpi_gtl_hsa"

Test Lightning-GPU with MPI
===========================

You can test the Python layer of the MPI enabled plugin as follows:

.. code-block:: bash

    mpirun -np 2 python -m pytest mpitests --tb=short

The C++ code can be tested with:

.. code-block:: bash

    PL_BACKEND="lightning_kokkos" make test-cpp-mpi

> Building and Running Lightning-Kokkos with MPI on Frontier

Author: This is probably not the page for this content - I think we agreed to have an extra page just for compiling and running Lightning-Kokkos on HPC. Where should I put it?

Reviewer: I think it could make sense to have this new page in the same place we keep the Overview, Installation, etc. pages, alternatively, in the 'Usage' section. cc @tomlqc.

Author: I have now created a new installation page

and exporting the install location to ``CMAKE_PREFIX_PATH`` as described above.
Then Lightning-Kokkos with MPI support can be installed in the *editable* mode by adding the `ENABLE_MPI=ON` option to the CMake arguments:

.. code-block:: bash

    git clone https://github.com/PennyLaneAI/pennylane-lightning.git
    cd pennylane-lightning
    PL_BACKEND="lightning_qubit" python scripts/configure_pyproject_toml.py
    SKIP_COMPILATION=True pip install -e . --config-settings editable_mode=compat
    PL_BACKEND="lightning_kokkos" python scripts/configure_pyproject_toml.py
    CMAKE_ARGS="-DENABLE_MPI=ON" python -m pip install -e . --config-settings editable_mode=compat -vv

If required, extra linker flags for MPI (e.g. for GPU Transport Layer) can be added using the ``MPI_EXTRA_LINKER_FLAGS`` environment variable, for example:

.. code-block:: bash

    export MPI_EXTRA_LINKER_FLAGS="-lxpmem -L/opt/cray/pe/mpich/8.1.31/gtl/lib -lmpi_gtl_hsa"

Test Lightning-Kokkos with MPI
===========================

You can test the Python layer of the MPI enabled plugin as follows:

.. code-block:: bash

>     mpirun -np 2 python -m pytest mpitests --tb=short

Reviewer 2: ```suggestion
    PL_BACKEND="lightning_kokkos" mpirun -np 2 python -m pytest mpitests --tb=short
```

Reviewer 2: or 
```suggestion
    make test-python-mpi device=lightning.kokkos
```

The C++ code can be tested with:

.. code-block:: bash

>     PL_BACKEND="lightning_kokkos" make test-cpp-mpi

Reviewer 2: ```suggestion
    PL_BACKEND="lightning_kokkos" make test-cpp-mpi build_options="-DKokkos_ENABLE_XYZ=ON"
```

Reviewer 2: or 
```suggestion
    CMAKE_ARGS="-DKokkos_ENABLE_OPENMP=ON" PL_BACKEND="lightning_kokkos" make test-cpp-mpi
```

Reviewer 2: And mentions that it is necessary to change the `XYZ`

Author: i recommend pre-compiling Kokkos beforehand and using `CMAKE_PREFIX_PATH`, instead of `build_options`. People should also add the appropriate MPI flags here too with `MPI_EXTRA_LINKER_FLAGS`. I'll update the instructions.

Building and Running Lightning-Kokkos with MPI on Frontier
==============================================

> To build Lightning-Kokkos with MPI on Frontier for AMD GPUs, the following commands can be used:

Reviewer 2: I suggest adding a link to the Frontier website 
https://www.olcf.ornl.gov/frontier/

Similarly, the CMake option ``-DKokkos_ARCH_{...}=ON`` must also be specified to target a given architecture.
A list of the architectures is found on the `Kokkos wiki <https://kokkos.org/kokkos-core-wiki/API/core/Macros.html#architectures>`_.
Note that ``THREADS`` backend is not recommended since `Kokkos does not guarantee its safety <https://github.com/kokkos/kokkos-core-wiki/blob/17f08a6483937c26e14ec3c93a2aa40e4ce081ce/docs/source/ProgrammingGuide/Initialization.md?plain=1#L67>`_.

.. _install-lightning-kokkos-with-mpi:

Install Lightning-Kokkos with MPI
=================================

.. note::

    Building Lightning-Kokkos with MPI requires a MPI library and ``mpi4py``. 

> To install Lightning-Kokkos with MPI support, we recommend first installing Kokkos for your specific architecture (CPU, Nvidia/AMD GPU etc.),

Reviewer 2: ```suggestion
To install Lightning-Kokkos with MPI support, we recommend first installing Kokkos for your specific architecture such as CPU (``SERIAL``, ``OPENMP``),  Nvidia GPU (``CUDA``), or AMD GPU (``HIP``)
```

and exporting the install location to ``CMAKE_PREFIX_PATH`` as described above.

> Then Lightning-Kokkos with MPI support can be installed in the *editable* mode by adding the `ENABLE_MPI=ON` option to the CMake arguments:

Reviewer 3: ```suggestion
Then Lightning-Kokkos with MPI support can be installed in the *editable* mode by adding the ``ENABLE_MPI=ON`` option to the CMake arguments:
```

.. code-block:: bash

    git clone https://github.com/PennyLaneAI/pennylane-lightning.git
    cd pennylane-lightning
    PL_BACKEND="lightning_qubit" python scripts/configure_pyproject_toml.py
    SKIP_COMPILATION=True pip install -e . --config-settings editable_mode=compat
    PL_BACKEND="lightning_kokkos" python scripts/configure_pyproject_toml.py
    CMAKE_ARGS="-DENABLE_MPI=ON" python -m pip install -e . --config-settings editable_mode=compat -vv

If required, extra linker flags for MPI (e.g. for GPU Transport Layer) can be added using the ``MPI_EXTRA_LINKER_FLAGS`` environment variable, for example:

> .. code-block:: bash
> 
>     export MPI_EXTRA_LINKER_FLAGS="-lxpmem -L/opt/cray/pe/mpich/8.1.31/gtl/lib -lmpi_gtl_hsa"

Reviewer 2: ```suggestion
.. code-block:: bash

    # Optional and system dependent
    export MPI_EXTRA_LINKER_FLAGS="-lxpmem -L/opt/cray/pe/mpich/8.1.31/gtl/lib -lmpi_gtl_hsa"
```

Similarly, the CMake option ``-DKokkos_ARCH_{...}=ON`` must also be specified to target a given architecture.
A list of the architectures is found on the `Kokkos wiki <https://kokkos.org/kokkos-core-wiki/API/core/Macros.html#architectures>`_.
Note that ``THREADS`` backend is not recommended since `Kokkos does not guarantee its safety <https://github.com/kokkos/kokkos-core-wiki/blob/17f08a6483937c26e14ec3c93a2aa40e4ce081ce/docs/source/ProgrammingGuide/Initialization.md?plain=1#L67>`_.

.. _install-lightning-kokkos-with-mpi:

Install Lightning-Kokkos with MPI
=================================

.. note::

>     Building Lightning-Kokkos with MPI requires a MPI library and ``mpi4py``. 

Reviewer 4: ```suggestion
    Building Lightning-Kokkos with MPI requires an MPI library and ``mpi4py``. 
```

To install Lightning-Kokkos with MPI support, we recommend first installing Kokkos for your specific architecture such as CPU (``SERIAL``, ``OPENMP``),  Nvidia GPU (``CUDA``), or AMD GPU (``HIP``)
and exporting the install location to ``CMAKE_PREFIX_PATH`` as described above.
Then Lightning-Kokkos with MPI support can be installed in the *editable* mode by adding the ``ENABLE_MPI=ON`` option to the CMake arguments:

.. code-block:: bash

    git clone https://github.com/PennyLaneAI/pennylane-lightning.git
    cd pennylane-lightning
    PL_BACKEND="lightning_qubit" python scripts/configure_pyproject_toml.py

>     SKIP_COMPILATION=True pip install -e . --config-settings editable_mode=compat

Reviewer 4: would you like to add a short description about this step?
